### PR TITLE
Simple proxy for tavern replacement enhanced

### PR DESCRIPTION
**Pull request by WolframRavenwolf**
I've continued working on the simple-proxy-for-tavern-replacement and enhanced it further:

- **Added Persona Description and System Prompt to Story String**

  Allows more flexible positioning of Persona Description and System Prompt inside the Story String. This is used by the enhanced simple-proxy-for-tavern context/instruct templates to exactly replicate simple-proxy-for-tavern's default output.

- **Added Include Names and Forced Names to all instruct presets**

  Since the enhanced simple-proxy-for-tavern instruct template works best without the normal names, I made those options part of the presets. All other presets got the default values.

### CODE REVIEW

File: public/script.js [hunk 1]
```diff
@@ -2527,9 +2525,11 @@ async function Generate(type, { automatic_trigger, force_name2, resolve, reject,
         const scenarioText = chat_metadata['scenario'] || characters[this_chid].scenario;
         let charDescription = baseChatReplace(characters[this_chid].description.trim(), name1, name2);
         let charPersonality = baseChatReplace(characters[this_chid].personality.trim(), name1, name2);
+        let personaDescription = baseChatReplace(power_user.persona_description.trim(), name1, name2);
         let Scenario = baseChatReplace(scenarioText.trim(), name1, name2);
         let mesExamples = baseChatReplace(characters[this_chid].mes_example.trim(), name1, name2);
         let systemPrompt = power_user.prefer_character_prompt ? baseChatReplace(characters[this_chid].data?.system_prompt?.trim(), name1, name2) : '';
+        systemPrompt = power_user.prefer_character_prompt && systemPrompt ? systemPrompt : baseChatReplace(power_user.instruct.system_prompt, name1, name2);
```
Comment (Cohee1207) [inline public/script.js]: This will cause an unintended side effect of using instruct system prompt as a system prompt for Chat Completions

Comment (Cohee1207) [reply public/script.js]: Easiest fix:
```
if (main_api !== 'openai')
```

File: public/context/Default.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "Default",
-    "story_string": "{{#if description}}{{description}}\n{{/if}}{{#if personality}}{{char}}'s personality: {{personality}}\n{{/if}}{{#if scenario}}Scenario: {{scenario}}{{/if}}",
+    "story_string": "{{#if system}}{{system}}\n{{/if}}{{#if persona}}{{persona}}\n{{/if}}{{#if description}}{{description}}\n{{/if}}{{#if personality}}{{char}}'s personality: {{personality}}\n{{/if}}{{#if scenario}}Scenario: {{scenario}}\n{{/if}}",
```
Comment (Cohee1207) [inline public/context/Default.json]: I think the after scenario would be a better default position for the persona as having it before may have a detrimental effect on attention to character description.

File: public/script.js (hunk 1 above)
Comment (WolframRavenwolf) [reply public/script.js]: OK, fixed: In "Instruct system prompt for instruct mode only", I put that inside an if clause that checks for Instruct mode.

File: public/context/Default.json (hunk 2 above)
Comment (WolframRavenwolf) [reply public/context/Default.json]: OK, done: I fully agree, I only put it there in the Default context to be the same as with the former default position "before character card". In "Position persona after scenario", I put it after scenario, same as the other contexts.

**Comment by WolframRavenwolf** [comment]
There's just one more thing missing from what the proxy does by default: It also changed the example messages, inserting the Input/Output sequences as would happen if the example messages were sent through the proxy. I'm currently working on that, so you might want to wait with merging until that's done - or merge now and I'll make a new PR for the example adjustments, up to you!

**Comment by Cohee1207** [comment]
I'm working with other PR right now, so you can do it here

**Comment by WolframRavenwolf** [comment]
**Format instruct mode examples:** Formats example messages, inserting Input/Output sequences, as if they were sent as normal chat messages.

**Llama 2 instruct mode preset fixed:** When testing example messages formatting with other instruct mode presets besides the proxy's, I noticed some differences from how it's supposed to look ([How to Prompt Llama 2](https://huggingface.co/blog/llama2#how-to-prompt-llama-2)), so I fixed them as best I could.

**Comment by WolframRavenwolf** [comment]
🥳 I think this is it! 🎉

We've finally achieved parity with simple-proxy-for-tavern's default verbose setup! I was comparing the outputs from proxy and preset and they're now 100 % the same (after turning on "Remove Empty New Lines from Output" in SillyTavern which the proxy does on its own).

And it works extremely well! In fact, I tested it with Aqua and Llama 2 13B Chat (yes, the censored model) and it uncensored it nicely. (Ironically, using the official Llama 2 prompt invokes the censorship, but the proxy preset works around that. 😏)

**Comment by Cohee1207** [comment]
Nice, we did it.
The next release is going to be 🔥 

Also, just a heads-up, I'm doing some minor refactorings for instruct mode, mainly reorganizing code and adding it to the preset manager. Also would need to come up with regex to smartly activate presets based on model names reported by the API. Then it would greatly simplify the experience for casual users. Will keep you posted.

**Comment by WolframRavenwolf** [comment]
Oh yeah, that will be a big one! 🚀

I'm focussing now on testing both proxy and Roleplay presets to find out if the Roleplay preset is a good-enough simplified/distilled version of it, or if the proxy is superior in all situations.

**If** the proxy is so much better, we could remove the Roleplay preset and possibly rename the proxy stuff to "Roleplay" for the next stable release - because "simple-proxy-for-tavern" is a strange name for a preset and meaningless for those who don't know what the proxy did. Having the proxy stuff as the new Roleplay preset would make it obvious what it's for.

But if the Roleplay preset proves to be useful, we should keep both. Proxy for purists who want 100 % the same as the proxy did, and Roleplay for those who want an evolved, streamlined version. After all, it saves a bunch of tokens and might work in situations where the proxy doesn't because of the different syntax, but I've got more testing to do before I can confirm that.

By the way: Since these presets work so well universally, I think it'd be best if one of them were the default instruct mode preset. Is the regex selection stuff optional? If not, I'd prefer if it was an option, because otherwise it might choose e. g. the official Llama 2 Chat preset and leave you with a censored model, while the proxy/Roleplay preset would uncensor it and work even better than the official format. The same is probably true for other formats, so while I welcome proper prompt formats for e. g. assistant use, if you want to roleplay, the generic proxy/Roleplay preset would actually work better than the "proper" format (as I've found out over months of testing with the proxy - I even submitted PRs for specific formats to the proxy, but then stuck to the defaults, because they worked so well in all situations).

**Comment by Cohee1207** [comment]
Regex is planned as a push toward selecting instruct preset and auto-enabling instruct mode for casual users who are clueless about the instruct mode. For now, I haven't actually put any of the regexes in, so it won't activate unless you type the regex in yourself. For the default preset selection (the one with the heart icon), it acts like a fallback if the instruct mode is enabled, but no presets were activated by the regex. The actual UX and logic could be tweaked in more detail when (and if) we get some feedback.

Some alternative considerations/speculations:
1. Always prefer the default preset, even if a more fitting model preset was activated by the regex, so it acts like a lock.
2. Never auto-enable instruct mode, it should be an informed decision. But allow auto-selection if was enabled manually.

**Comment by WolframRavenwolf** [comment]
Regarding those alternatives:

1. When I saw the "Default" button, I assumed it to be a lock, and wondered why I couldn't toggle it off. So personally, I'd really prefer that as a toggle button, and when it's enabled, I'd want this preset to be used all the time. Because I know I'd always want to use the proxy/Roleplay preset, no matter the model, and only select a different one in rare special cases. I'm convinced it works so well that from a newbie perspective, it would be best if that was the default, and power users could switch to another manually or automatically selected preset if they want.

2. I'm not sure why anyone would not want instruct mode, because it gives so much control through the prompt options. However, I might look at it differently because I've always used the proxy or Roleplay preset, which is independent of if it's a chat or instruct model and its prompt format - it just works. So that's why I've always been a big fan of instruct mode.

Another suggestion for more ease-of-use:

Right now, it's three clicks to activate the proxy/Roleplay presets: Select the Context template, select the Instruct mode preset, and make sure Instruct mode is enabled. So linking Context template and Instruct mode preset together (e. g. by the logic that if you select one, and there's another with the same name, it would also select that) would be a nice quality of life improvement.

Maybe even replace the Enabled checkbox with just the preset selection and offer "Disabled" as an option inside the drop-down list. Would simplify the UI without losing functionality. I'd also expect the preset selection drop-down above the checkboxes that are modified by it. (I was considering doing a PR for that but don't want to interfere with what you're doing currently.)

**Comment by Cohee1207** [comment]
This is already pushed to staging but with instruct auto-enabling removed. Would think about alternative nudges toward instruct mode enabling rather than doing it without user consent. "Default" is now a toggle, it was a placeholder control the day before. Will probably swap the order of operations and do process the default preset first. By the way, this logic kicks in only on connection to API when the model name is known (we just don't talk about koboldcpp).

**Comment by Cohee1207** [comment]
Also, you're free to do any improvements now, the refactor is finished (check out the new file in https://github.com/SillyTavern/SillyTavern/blob/staging/public/scripts/instruct-mode.js )

**Pull request merged**